src/mip/coefficient_strengthening.jl

```diff
@@ -0,0 +1,169 @@
+aᵢ x ⩽ bᵢ and let xⱼ be a integer/binary variable
+If aᵢⱼ ⩾ d = bᵢ - Mᵢⱼ - aᵢⱼ (uⱼ - 1) > 0
+where Mᵢⱼ is maximal activity of the i-th row without considering xⱼ
+and uⱼ is upper bound of xⱼ, then
+aᵢⱼ ≔ aᵢⱼ - d
+bᵢ ≔ bᵢ - duⱼ
+Also perform coefficient strengthening on constraints of the form aᵢ x ⩾ cᵢ
+with similar update rule
+
+Skip ranged constraints, i.e, cᵢ ⩽ aᵢ x ⩽ bᵢ
+
+The new coefficients are stored by updating ps.pb0
+If some coefficient is reduced to 0, it is kept as 0
+in both ps.pb0.arows and ps.pb0.acols, and only the nonzeros are updated
+
+If a row is reduced to 0 after this step, it is set to be inactive
+If a column is reduced to 0 after this step, it is set to be inactive
+
+"""
+function maximal_activity(row::Row{T}, ucol::Vector{T}, lcol::Vector{T})::T where {T}
+    sup = zero(T)
+
+    for (j, aij) in zip(row.nzind, row.nzval)
+        if aij > zero(T)
+            sup += aij*ucol[j]
+        else
+            sup += aij*lcol[j]
+        end
+    end
+    return T(sup)
+end
+
+function minimal_activity(row::Row{T}, ucol::Vector{T}, lcol::Vector{T})::T where {T}
+    inf = zero(T)
+
+    for (j, aij) in zip(row.nzind, row.nzval)
+        if aij > zero(T)
+            inf += aij*lcol[j]
+        else
+            inf += aij*ucol[j]
+        end
+    end
+    return T(inf)
+end
+
+function upperbound_strengthening!(ps::PresolveData{T}, i::Int, j_index::Int, j::Int, max_act) where {T}
+    # perform coef strengthening for one constraints of the from a'x <= u
+    row = ps.pb0.arows[i]
+    a = row.nzval
+    u = ps.urow[i]
+    if a[j_index] > 0
+        d = u - max_act - a[j_index]*(ps.ucol[j]-1)
+        if a[j_index] >= d > 0
+            a[j_index] = a[j_index] - d
+            ps.urow[i] = u - d*ps.ucol[j]
+        end
+    elseif a[j_index] < 0
+        d = u - max_act - a[j_index]*(ps.lcol[j]+1)
+        if -a[j_index] >= d > 0
+            a[j_index] = a[j_index] + d
+            ps.urow[i] = u + d*ps.lcol[j]
+        end
+    end
+    return T(a[j_index])
+end
+
+function lowerbound_strengthening!(ps::PresolveData{T}, i::Int, j_index::Int, j::Int, min_act) where {T}
+    # perform coef strengthening for one constraints of the from l < = a'x
+    row = ps.pb0.arows[i]
+    a = row.nzval
+    l = ps.lrow[i]
+    if a[j_index] > 0
+        d = -l + min_act + a[j_index]*(ps.lcol[j]+1)
+        if a[j_index] >= d > 0
+            a[j_index] = a[j_index] - d
+            ps.lrow[i] = l - d*ps.lcol[j]
+        end
+    elseif a[j_index] < 0
+        d = -l - min_act + a[j_index]*(ps.ucol[j]-1)
+        if -a[j_index] >= d > 0
+            a[j_index] = a[j_index] + d
+            ps.urow[i] = u + d*ps.ucol[j]
+        end
+    end
+    return T(a[j_index])
+end
+
+function zero_coefficient_strengthening!(ps::PresolveData{T}) where {T}
+    # perform coefficient stregthening but if there is a coefficient is reduced to 0
+    # it is still kept in the ps.pb0.arows and ps.pb0.acols
```

Comment:
cc @joehuchette 
How do we want to handle coefficients that become zero?
Do we want to systematically remove them, or keep "numerical" zeros in the data structures?

If it's the former, we need to do the proper book-keeping when removing them.
If it's the latter, we need to ensure that we never do something like `0 * Inf`, and that we correctly keep track of per-row and per-column non-zeros.

Reply:
That's also the question I have about how to handle coefficient that becomes zero.
I checked your extract_reduced_problem function, and it seems to me that you read the coefficient from ps.pb0, and having rowflag & colflag to keep track of inactive rows and cols, so for now, I handle them by keep "numerical" zeros in the data structures.

Reply:
Let's keep them for now then. We just need to keep in mind that some coefficients may be zero :)

Reply:
I think we should have a separate routine that eliminates all explicitly stored zeros in one pass.